pkg/auth/requests/impersonate.go

```diff
@@ -55,36 +57,47 @@ func (h *impersonatingAuth) Authenticate(req *http.Request) (k8sUser.Info, bool,
 	// impersonate a different user, verify the token user is authz to impersonate
 	if h.sar != nil {
 		if reqUser != "" && reqUser != user {
-			canDo, err := h.sar.UserCanImpersonateUser(req, user, reqUser)
-			if err != nil {
-				return nil, false, err
-			} else if !canDo {
-				return nil, false, errors.New("not allowed to impersonate user")
-			}
-			impersonateUser = true
-
-			for _, g := range reqGroup {
-				if slices.Contains(groups, g) {
-					//user belongs to the group they are trying to impersonate
-					continue
-				}
-				canDo, err := h.sar.UserCanImpersonateGroup(req, user, g)
+			if strings.HasPrefix(reqUser, serviceaccount.ServiceAccountUsernamePrefix) {
+				canDo, err := h.sar.UserCanImpersonateServiceAccount(req, user, reqUser)
 				if err != nil {
-					return nil, false, fmt.Errorf("error checking if user can impersonate group: %w", err)
+					return nil, false, err
 				} else if !canDo {
-					return nil, false, errors.New("not allowed to impersonate group")
+					return nil, false, errors.New("not allowed to impersonate service account")
 				}
-				impersonateGroup = true
-			}
-
-			if len(reqExtras) > 0 {
-				canDo, err := h.sar.UserCanImpersonateExtras(req, user, reqExtras)
+				// add impersonated SA to context
+				*req = *req.WithContext(authcontext.SetSAImpersonation(req.Context(), reqUser))
+			} else {
+				canDo, err := h.sar.UserCanImpersonateUser(req, user, reqUser)
 				if err != nil {
-					return nil, false, fmt.Errorf("error checking if user can impersonate extras: %w", err)
+					return nil, false, err
```

Comment:
Don't we want to wrap the error here?
```suggestion
					return nil, false, fmt.Errorf("error checking if user can impersonate another user: %w", err)
```